KAFKA-18913: Start state updater in task manager

janchilling opened the pull request:
Updated the code to start the State Updater Thread only after the Stream
Thread is started.

Changes done :
1. Moved the starting of the StateUpdater thread to a new init method in
the TaskManager.
2. Called the init of TaskManager in the run method of the StreamThread.
3. Updated the test cases in the StreamThreadTest to mimic the
aforementioned behaviour.

Reviewers: Bruno Cadonna <cadonna@apache.org>

janchilling committed:
KAFKA-18913: State Updater thread not shutting down during teardown in StreamThreadTest fix

Updated the code to start the State Updater Thread only after the Stream Thread is started.

Changes done :
1. Moved the starting of the StateUpdater thread to a new init method in the TaskManager.
2. Called the init of TaskManager in the run method of the StreamThread.
3. Updated the test cases in the StreamThreadTest to mimic the aforementioned behaviour.

janchilling commented:
Hi @cadonna ,

Here is the fix for the issue that we discussed(`StateUpdater` thread being left to run even though an IllegalStateException thrown before the `StreamThread` is created).
I hope this is the implementation that you requested to be done. If there are any changes or if anything needs to be added please let me know. I might be away for the next 2 weeks, I'll attend to any required changes when I find time.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -149,6 +149,11 @@ public class TaskManager {
         );
     }
 
+    void init() {
+        if (stateUpdater != null) {
+            this.stateUpdater.start();
+        }
+    }
```
Could you please add a unit test that verifies that the state updater is indeed started when the task manager is initialized?

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -4117,6 +4149,12 @@ private StreamThread buildStreamThread(final Consumer<byte[], byte[]> consumer,
+        if (stateUpdaterEnabled) {
+            thread.taskManager().init();
+        }
+    }
```
It should be enough to call `thread.taskManager().init()` instead of this method. Whether the state updater is enabled is already checked in the `init()`.

You also do not need to call it in each and every test method. Only in some the call is needed.

janchilling replied:
Hi @cadonna ,

I actually didn't call it in every test method, only in the ones where the StreamThread is created by using the `StreamThreadTest#createStreamThread`. Therefore, to keep the behavior uniform with the code, I included it in those test cases.  

That said, if the above approach isn’t ideal, another option could be to include the call in the tests where tasks are not Restoring properly. I tried this alternative as well, and all test cases passed with this approach as well.

Happy to adjust it either way, depending on what you think would be best.

cadonna replied:
Fair enough!
I checked a couple of test methods and found that in at least two cases `maybeRunStateUpdater()` was not needed. The two cases are: `shouldRecordCommitLatency()` and `shouldChangeStateInRebalanceListener()`. There might still be others.

cadonna reviewed (commented):
@janchilling Thanks for the PR!

The PR corresponds to my proposal in the other PR.

I left some comments.

cadonna commented:
> 
> I might be away for the next 2 weeks, I'll attend to any required changes when I find time.

No worries! Thanks for letting us know!

janchilling committed:
KAFKA-18913: State Updater thread not shutting down during teardown in StreamThreadTest fix

Added a test case to verify whether the StateUpdater thread is running after calling the init method in the TaskManager and to check if it shuts down properly.

Furthermore, did some changes requested in the review in the StreamThreadTest class

janchilling commented:
Hi @cadonna ,

I’ve added a test case that verifies both the starting and shutdown behavior of the `StateUpdater` thread—since we shouldn't leave the thread running after the test.

To perform the verification, I used `DefaultStateUpdater` directly in order to access the `isRunning()` method, as the existing mock `StateUpdater` doesn't expose it. This meant writing a more customized test rather than reusing the existing setup methods.

I believe this addresses all the requested changes, but please let me know if there’s anything else you'd like me to update.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -4837,6 +4841,44 @@ public void shouldUseStartupTasksFromStateDirectoryAsStandbyWithStateUpdater() {
+        );
+
+        taskManager.init();
+        assertTrue(defaultStateUpdater.isRunning());
```
You do not need to check for `isRunning()` here. `isRunning()` is an implementation detail of the `DefaultStateUpdater`. It suffices to verify that `stateUpdater.start()` is called on the mock. Something simple like the following:

```java
    @ParameterizedTest
    @ValueSource(booleans = {true, false})
    public void shouldStartStateUpdaterOnInit(final boolean stateUpdaterEnabled) {
        final TaskManager taskManager = setUpTaskManager(ProcessingMode.AT_LEAST_ONCE, stateUpdaterEnabled);
        taskManager.init();
        if (stateUpdaterEnabled) {
            verify(stateUpdater).start();
        } else {
            verify(stateUpdater, never()).start();
        }
    }
```
The shutdown is tested elsewhere. You do not need to test it again here.

janchilling replied:
Thanks for pointing that out.  I've updated the test accordingly.

cadonna reviewed (commented):
@janchilling Thanks for the update!

Here my comment.

Almost there.

janchilling committed:
KAFKA-18913: State Updater thread not shutting down during teardown in StreamThreadTest fix

Test case added to verify that the stateUpdater.start() is called on the mock StateUpdater.

cadonna reviewed (approved):
Thanks @janchilling!

LGTM!
